lib/buffer.js

```diff
@@ -24,82 +24,148 @@ function createPool() {
 }
 createPool();
 
+function Buffer(arg) {
+  if (!(this instanceof Buffer)) {
+    // Avoid going through an ArgumentsAdaptorTrampoline in the common case.
+    if (arguments.length > 1)
+      return new Buffer(arg, arguments[1]);
 
-function Buffer(subject, encoding) {
-  if (!(this instanceof Buffer))
-    return new Buffer(subject, encoding);
-
-  if (typeof subject === 'number') {
-    this.length = +subject;
-
-  } else if (typeof subject === 'string') {
-    if (typeof encoding !== 'string' || encoding.length === 0)
-      encoding = 'utf8';
-    this.length = Buffer.byteLength(subject, encoding);
+    return new Buffer(arg);
+  }
 
-  // Handle Arrays, Buffers, Uint8Arrays or JSON.
-  } else if (subject !== null && typeof subject === 'object') {
-    if (subject.type === 'Buffer' && Array.isArray(subject.data))
-      subject = subject.data;
-    this.length = +subject.length;
+  this.length = 0;
+  this.parent = undefined;
 
-  } else {
-    throw new TypeError('must start with number, buffer, array or string');
+  // Common case.
+  if (typeof(arg) === 'number') {
+    fromNumber(this, arg);
+    return;
   }
 
-  if (this.length > kMaxLength) {
-    throw new RangeError('Attempt to allocate Buffer larger than maximum ' +
-                         'size: 0x' + kMaxLength.toString(16) + ' bytes');
+  // Slightly less common case.
+  if (typeof(arg) === 'string') {
+    fromString(this, arg, arguments.length > 1 ? arguments[1] : 'utf8');
+    return;
   }
 
-  if (this.length < 0)
-    this.length = 0;
-  else
-    this.length >>>= 0;  // Coerce to uint32.
+  // Unusual.
+  fromObject(this, arg);
+}
 
-  this.parent = undefined;
-  if (this.length <= (Buffer.poolSize >>> 1) && this.length > 0) {
-    if (this.length > poolSize - poolOffset)
-      createPool();
-    this.parent = sliceOnto(allocPool,
-                            this,
-                            poolOffset,
-                            poolOffset + this.length);
-    poolOffset += this.length;
-  } else {
-    alloc(this, this.length);
-  }
+function fromNumber(that, length) {
+  allocate(that, length < 0 ? 0 : checked(length) >>> 0);
```

Comment:
It's supposed to be a strong hint to V8 that the result is an int but I checked it again and it doesn't seem to be effective...  I'll replace them with `checked(length) | 0`, that does eliminate the DoubleToIStub calls.